File: packages/android/src/main/java/com/kozmos/components/ManoeuvreCard/ManoeuvreCard.kt
package com.kozmos.components.manoeuvrecard

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.collapse
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.expand
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.kozmos.components.directionstep.DirectionType
import com.kozmos.components.directionstep.icon
import com.kozmos.tokens.KozmosColors
import com.kozmos.tokens.KozmosDimensions

/** What TalkBack hears for the closed card: the instruction, then the detail. */
fun manoeuvreDescription(instruction: String, detail: String?): String =
    if (detail.isNullOrEmpty()) instruction else "$instruction, $detail"

/**
 * The current manoeuvre, floating over the map during navigation: its arrow,
 * the instruction, how far and how long, and a grab bar that opens the full
 * itinerary in its place. The card owns the toggle and what TalkBack hears
 * of it; the itinerary it opens into is the caller's — `KozmosItinerary`, in
 * the products — so the card never decides what a route is made of. Open,
 * the card is as tall as the itinerary up to `maxItineraryHeight`, past
 * which the itinerary scrolls: a long route must not cover the map.
 */
@Composable
fun KozmosManoeuvreCard(
    type: DirectionType,
    instruction: String,
    expanded: Boolean,
    onToggle: () -> Unit,
    modifier: Modifier = Modifier,
    detail: String? = null,
    expandLabel: String = "Show itinerary",
    collapseLabel: String = "Hide itinerary",
    manoeuvreLabel: String = "Current manoeuvre",
    maxItineraryHeight: Dp = 320.dp,
    itinerary: @Composable () -> Unit
) {
    Surface(
        // Open, the card has no name of its own: the itinerary inside is the
        // named thing, and two nodes called the same would be read twice.
        modifier = modifier.then(if (expanded) Modifier else Modifier.semantics { contentDescription = manoeuvreLabel }),
        shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusContainer),
        color = KozmosColors.primitivesColorsBackground0.copy(alpha = 0.9f),
        shadowElevation = 8.dp
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(
                    start = KozmosDimensions.primitivesLayoutSpacing200,
                    top = KozmosDimensions.primitivesLayoutSpacing200,
                    end = KozmosDimensions.primitivesLayoutSpacing200,
                    bottom = KozmosDimensions.primitivesLayoutSpacing50
                )
        ) {
            if (expanded) {
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .heightIn(max = maxItineraryHeight)
                        .verticalScroll(rememberScrollState())
                ) {
                    itinerary()
                }
            } else {
                // The instruction row is the button: a tap anywhere on it opens
                // the itinerary, and TalkBack hears the manoeuvre with the action.
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable(onClickLabel = expandLabel, role = Role.Button, onClick = onToggle)
                        .semantics(mergeDescendants = true) {
                            contentDescription = manoeuvreDescription(instruction, detail)
                            expand { onToggle(); true }
                        },
                    verticalAlignment = Alignment.Top
                ) {
                    Box(
                        modifier = Modifier.size(KozmosDimensions.primitivesLayoutSizing400),
                        contentAlignment = Alignment.Center
                    ) {
                        Icon(
                            imageVector = type.icon(),
                            contentDescription = null,
                            tint = KozmosColors.primitivesColorsTheme500,
                            modifier = Modifier.size(KozmosDimensions.primitivesLayoutSizing300)
                        )
                    }
                    Column(modifier = Modifier.weight(1f).padding(start = KozmosDimensions.primitivesLayoutSpacing150)) {
                        Text(
                            text = instruction,
                            style = MaterialTheme.typography.titleLarge.copy(fontWeight = FontWeight.SemiBold),
                            color = KozmosColors.primitivesColorsForeground100,
                            maxLines = 2,
                            overflow = TextOverflow.Ellipsis
                        )
                        if (!detail.isNullOrEmpty()) {
                            Text(
                                text = detail,
                                style = MaterialTheme.typography.bodyMedium,
                                color = KozmosColors.primitivesColorsForeground500
                            )
                        }
                    }
                }
            }
            // The grab bar: the sign that the card opens, and the way to close
            // it. Closed, the instruction row already offers the way in.
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = KozmosDimensions.primitivesLayoutSpacing150)
                    // Closed, the bar has no semantics at all — the cleared
                    // node hides the click that follows it — so TalkBack
                    // never lands on a second way to do what the row does.
                    .then(
                        if (expanded) {
                            Modifier.semantics {
                                contentDescription = collapseLabel
                                collapse { onToggle(); true }
                            }
                        } else {
                            Modifier.clearAndSetSemantics { }
                        }
                    )
                    .clickable(onClick = onToggle, role = Role.Button)
                    .padding(vertical = KozmosDimensions.primitivesLayoutSpacing50),
                contentAlignment = Alignment.Center
            ) {
                Box(
                    modifier = Modifier
                        .width(36.dp)
                        .height(5.dp)
                        .background(KozmosColors.primitivesColorsBackground300, CircleShape)
                )
            }
        }
    }
}

File: packages/ios/Sources/Components/ManoeuvreCard/ManoeuvreCard.swift
import SwiftUI

/// The current manoeuvre, floating over the map during navigation: its arrow,
/// the instruction, how far and how long, and a grab bar that opens the full
/// itinerary in its place. Mirrors the product prototype's instruction card:
/// 402 wide it is 378 × 119, radius 18, white at 90 %, the instruction 20/600
/// over a 14 grey detail, a 36 × 5 grab bar at its foot.
///
/// The card owns the toggle and what VoiceOver hears of it. The itinerary it
/// opens into is the caller's — `KozmosItinerary`, in the products — so the
/// card never decides what a route is made of. Open, the card is as tall as
/// the itinerary up to `maxItineraryHeight`, past which the itinerary scrolls:
/// a long route must not cover the map.
public struct KozmosManoeuvreCard<Itinerary: View>: View {
    let type: DirectionType
    let instruction: String
    let detail: String?
    let isExpanded: Bool
    let onToggle: () -> Void
    let expandLabel: String
    let collapseLabel: String
    let maxItineraryHeight: CGFloat
    let itinerary: Itinerary

    public init(
        type: DirectionType,
        instruction: String,
        detail: String? = nil,
        isExpanded: Bool,
        onToggle: @escaping () -> Void,
        expandLabel: String = "Show itinerary",
        collapseLabel: String = "Hide itinerary",
        maxItineraryHeight: CGFloat = 320,
        @ViewBuilder itinerary: () -> Itinerary
    ) {
        self.type = type
        self.instruction = instruction
        self.detail = detail
        self.isExpanded = isExpanded
        self.onToggle = onToggle
        self.expandLabel = expandLabel
        self.collapseLabel = collapseLabel
        self.maxItineraryHeight = maxItineraryHeight
        self.itinerary = itinerary()
    }

    /// What VoiceOver reads for the closed card: the instruction, then the
    /// detail. The arrow says nothing the instruction does not.
    static func accessibilityDescription(instruction: String, detail: String?) -> String {
        [instruction, detail].compactMap { $0 }.filter { !$0.isEmpty }.joined(separator: ", ")
    }

    private var shape: RoundedRectangle {
        RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusContainer, style: .continuous)
    }

    public var body: some View {
        VStack(spacing: KozmosDimensions.primitivesLayoutSpacing150) {
            if isExpanded {
                // The itinerary as it is when it fits the cap, in a scroll
                // view when it does not: decided in one layout pass, so the
                // card never shows a frame at the wrong height. A scroll view
                // alone would take the whole cap for a route of three steps,
                // and so would `.frame(maxHeight:)`, which takes all of the
                // proposal up to its maximum; the layout below proposes the
                // cap and is only as tall as what it holds.
                KozmosCappedHeightLayout(cap: maxItineraryHeight) {
                    ViewThatFits(in: .vertical) {
                        itinerary
                            .frame(maxWidth: .infinity, alignment: .leading)
                        ScrollView {
                            itinerary
                                .frame(maxWidth: .infinity, alignment: .leading)
                        }
                    }
                }
            } else {
                // The instruction row is the button: a tap anywhere on it opens
                // the itinerary, and VoiceOver hears the manoeuvre with that hint.
                Button(action: onToggle) {
                    HStack(alignment: .top, spacing: KozmosDimensions.primitivesLayoutSpacing150) {
                        Image(systemName: type.iconName)
                            .font(.system(size: 22, weight: .semibold))
                            .foregroundColor(KozmosColors.primitivesColorsTheme500)
                            .frame(width: KozmosDimensions.primitivesLayoutSizing400, height: KozmosDimensions.primitivesLayoutSizing400)
                        VStack(alignment: .leading, spacing: KozmosDimensions.primitivesLayoutSpacing25) {
                            Text(instruction)
                                .font(KozmosTypography.title3.weight(.semibold))
                                .foregroundColor(KozmosColors.primitivesColorsForeground100)
                                .lineLimit(2)
                                .fixedSize(horizontal: false, vertical: true)
                            if let detail, !detail.isEmpty {
                                Text(detail)
                                    .font(KozmosTypography.subheadline)
                                    .foregroundColor(KozmosColors.primitivesColorsForeground500)
                            }
                        }
                        .frame(maxWidth: .infinity, alignment: .leading)
                    }
                    .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
                .accessibilityElement(children: .ignore)
                .accessibilityLabel(Self.accessibilityDescription(instruction: instruction, detail: detail))
                .accessibilityHint(expandLabel)
                .accessibilityAddTraits(.isButton)
            }
            // The grab bar: the sign that the card opens, and the way to close it.
            Button(action: onToggle) {
                Capsule()
                    .fill(KozmosColors.primitivesColorsBackground300)
                    .frame(width: 36, height: 5)
                    .padding(.vertical, KozmosDimensions.primitivesLayoutSpacing50)
                    .frame(maxWidth: .infinity)
                    .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel(isExpanded ? collapseLabel : expandLabel)
            // Closed, the instruction row already offers the way in.
            .accessibilityHidden(!isExpanded)
        }
        .padding(.top, KozmosDimensions.primitivesLayoutSpacing200)
        .padding(.horizontal, KozmosDimensions.primitivesLayoutSpacing200)
        .padding(.bottom, KozmosDimensions.primitivesLayoutSpacing50)
        .frame(maxWidth: .infinity)
        .background(KozmosColors.primitivesColorsBackground0.opacity(0.9))
        .clipShape(shape)
        .kozmosElevation(KozmosShadows.semanticsElevationFloating)
        .accessibilityElement(children: .contain)
        .accessibilityLabel(isExpanded ? "Itinerary" : "Current manoeuvre")
    }
}

/// Proposes at most `cap` to its one child and takes the child's size: the
/// child decides whether it fits, and the layout never takes room the child
/// does not fill. Where no height is proposed at all, the cap is.
struct KozmosCappedHeightLayout: Layout {
    let cap: CGFloat

    private func proposal(_ proposal: ProposedViewSize) -> ProposedViewSize {
        ProposedViewSize(width: proposal.width, height: min(proposal.height ?? cap, cap))
    }

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        subviews.first?.sizeThatFits(self.proposal(proposal)) ?? .zero
    }

    func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) {
        subviews.first?.place(at: bounds.origin, anchor: .topLeading, proposal: self.proposal(proposal))
    }
}
